File: notebooks/ed_length_of_stay_analysis.py
"""Emergency Department Length of Stay Analysis (Python)

Python conversion of the notebook workflow in:
- notebooks/ed_length_of_stay_analysis.ipynb

Run from repository root:
    python notebooks/ed_length_of_stay_analysis.py
"""

from __future__ import annotations

from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf
from scipy import stats
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split


ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "raw" / "ed_length_of_stay_100k.csv"
FIGURES_DIR = ROOT / "figures"
RANDOM_STATE = 123


def load_and_clean_data(file_path: Path) -> pd.DataFrame:
    df = pd.read_csv(file_path)

    df["vdate"] = pd.to_datetime(df["vdate"], format="%m/%d/%Y", errors="coerce")
    df["discharged"] = pd.to_datetime(df["discharged"], format="%m/%d/%Y", errors="coerce")

    df["gender"] = df["gender"].astype("category")
    rcount_levels = ["0", "1", "2", "3", "4", "5+"]
    df["rcount"] = pd.Categorical(df["rcount"].astype(str), categories=rcount_levels, ordered=True)
    df["facid"] = df["facid"].astype("category")
    df["lengthofstay"] = pd.to_numeric(df["lengthofstay"], errors="coerce")

    df = df[(df["sodium"] >= 110) & (df["sodium"] <= 170)]
    df = df[(df["neutrophils"] >= 0) & (df["neutrophils"] <= 100)]
    df = df[df["glucose"] >= 0]
    df = df[(df["bloodureanitro"] >= 0) & (df["bloodureanitro"] <= 200)]
    df = df[(df["respiration"] >= 4) & (df["respiration"] <= 40)]
    df = df[(df["hematocrit"] >= 5) & (df["hematocrit"] <= 60)]
    df = df[df["lengthofstay"] > 0]

    return df.dropna(subset=["lengthofstay"]).copy()


def run_eda(df: pd.DataFrame) -> None:
    FIGURES_DIR.mkdir(parents=True, exist_ok=True)

    print("\n=== EDA ===")
    print(df.describe(include="all").T.head(20))

    plt.figure(figsize=(8, 5))
    plt.hist(df["lengthofstay"], bins=20, edgecolor="black")
    plt.title("Histogram of Length of Stay")
    plt.xlabel("Length of Stay")
    plt.ylabel("Count")
    plt.tight_layout()
    plt.savefig(FIGURES_DIR / "los_histogram_python.png", dpi=140)
    plt.close()

    plt.figure(figsize=(6, 4))
    df["gender"].value_counts(dropna=False).plot(kind="bar")
    plt.title("Gender Distribution")
    plt.xlabel("Gender")
    plt.ylabel("Count")
    plt.tight_layout()
    plt.savefig(FIGURES_DIR / "gender_distribution_python.png", dpi=140)
    plt.close()

    plt.figure(figsize=(9, 4))
    df["facid"].value_counts(dropna=False).sort_index().plot(kind="bar")
    plt.title("Facility Distribution")
    plt.xlabel("Facility ID")
    plt.ylabel("Count")
    plt.tight_layout()
    plt.savefig(FIGURES_DIR / "facility_distribution_python.png", dpi=140)
    plt.close()


def run_inference(df: pd.DataFrame) -> None:
    print("\n=== Inference ===")

    male_los = df.loc[df["gender"] == "M", "lengthofstay"].dropna()
    female_los = df.loc[df["gender"] == "F", "lengthofstay"].dropna()
    t_stat, p_value = stats.ttest_ind(male_los, female_los, equal_var=False)
    print(f"Two-sample t-test (Gender): t = {t_stat:.4f}, p = {p_value:.6g}")

    anova_model = smf.ols("lengthofstay ~ C(facid)", data=df).fit()
    anova_table = sm.stats.anova_lm(anova_model, typ=2)
    print("\nOne-way ANOVA (Facility):")
    print(anova_table)


def get_formulas() -> dict[str, str]:
    full_formula = (
        "lengthofstay ~ C(gender) + C(rcount) + C(facid) + "
        "dialysisrenalendstage + asthma + irondef + pneum + "
        "substancedependence + psychologicaldisordermajor + depress + psychother + "
        "fibrosisandother + malnutrition + hemo + hematocrit + neutrophils + sodium + "
        "glucose + bloodureanitro + creatinine + bmi + pulse + respiration + "
        "secondarydiagnosisnonicd9"
    )

    clinical_formula = (
        "lengthofstay ~ C(rcount) + dialysisrenalendstage + pneum + "
        "psychologicaldisordermajor + depress + malnutrition + hematocrit + "
        "neutrophils + sodium + glucose + bloodureanitro + creatinine + bmi + "
        "pulse + respiration"
    )

    minimal_formula = "lengthofstay ~ C(rcount) + hematocrit + bloodureanitro + creatinine"

    return {
        "full": full_formula,
        "clinical": clinical_formula,
        "minimal": minimal_formula,
    }


def run_regression_workflow(df: pd.DataFrame) -> None:
    formulas = get_formulas()

    print("\n=== Baseline Regression (MLR) ===")
    mlr = smf.ols(formulas["full"], data=df).fit()
    print(mlr.summary().tables[0])
    print(mlr.summary().tables[1])

    print("\nAIC / BIC (nested models)")
    m_full = smf.ols(formulas["full"], data=df).fit()
    m_clinical = smf.ols(formulas["clinical"], data=df).fit()
    m_minimal = smf.ols(formulas["minimal"], data=df).fit()
    print(
        pd.DataFrame(
            {
                "Model": ["Full", "Clinical", "Minimal"],
                "AIC": [m_full.aic, m_clinical.aic, m_minimal.aic],
                "BIC": [m_full.bic, m_clinical.bic, m_minimal.bic],
            }
        )
    )

    train, test = train_test_split(df, test_size=0.30, random_state=RANDOM_STATE)

    m_full_tr = smf.ols(formulas["full"], data=train).fit()
    m_clinical_tr = smf.ols(formulas["clinical"], data=train).fit()
    m_minimal_tr = smf.ols(formulas["minimal"], data=train).fit()

    y_test = test["lengthofstay"]
    pred_full = m_full_tr.predict(test)
    pred_clinical = m_clinical_tr.predict(test)
    pred_minimal = m_minimal_tr.predict(test)

    mspe_results = pd.DataFrame(
        {
            "Model": ["Full", "Clinical", "Minimal"],
            "MSPE": [
                mean_squared_error(y_test, pred_full),
                mean_squared_error(y_test, pred_clinical),
                mean_squared_error(y_test, pred_minimal),
            ],
        }
    )

    print("\nMSPE Comparison")
    print(mspe_results)

    run_gamma_glm(df, train, test, formulas["clinical"], mspe_results)
    plot_mlr_diagnostics(mlr)


def run_gamma_glm(
    df: pd.DataFrame,
    train: pd.DataFrame,
    test: pd.DataFrame,
    glm_formula: str,
    mspe_results: pd.DataFrame,
) -> None:
    print("\n=== Gamma GLM (Improved Model) ===")

    gamma_family = sm.families.Gamma(sm.families.links.Log())
    glm_improved = smf.glm(glm_formula, data=df, family=gamma_family).fit()

    print(glm_improved.summary().tables[0])
    print(glm_improved.summary().tables[1])

    pseudo_r2 = 1 - (glm_improved.deviance / glm_improved.null_deviance)
    print(f"Pseudo R^2: {pseudo_r2:.4f}")
    print(f"AIC: {glm_improved.aic:.2f}")

    glm_improved_tr = smf.glm(glm_formula, data=train, family=gamma_family).fit()
    pred_glm = glm_improved_tr.predict(test)
    mspe_glm = mean_squared_error(test["lengthofstay"], pred_glm)

    mspe_all = pd.concat(
        [
            mspe_results,
            pd.DataFrame({"Model": ["Gamma GLM (improved)"], "MSPE": [mspe_glm]}),
        ],
        ignore_index=True,
    )

    print("\nMSPE Including Gamma GLM")
    print(mspe_all)

    plot_glm_diagnostics(glm_improved)


def plot_mlr_diagnostics(model: sm.regression.linear_model.RegressionResultsWrapper) -> None:
    FIGURES_DIR.mkdir(parents=True, exist_ok=True)

    residuals = model.resid
    fitted = model.fittedvalues

    plt.figure(figsize=(8, 5))
    plt.scatter(fitted, residuals, alpha=0.2)
    plt.axhline(0, linestyle="--")
    plt.title("MLR Residuals vs Fitted")
    plt.xlabel("Fitted values")
    plt.ylabel("Residuals")
    plt.tight_layout()
    plt.savefig(FIGURES_DIR / "mlr_residuals_vs_fitted_python.png", dpi=140)
    plt.close()

    plt.figure(figsize=(8, 5))
    stats.probplot(residuals, dist="norm", plot=plt)
    plt.title("MLR Normal Q-Q")
    plt.tight_layout()
    plt.savefig(FIGURES_DIR / "mlr_qq_python.png", dpi=140)
    plt.close()


def plot_glm_diagnostics(model: sm.genmod.generalized_linear_model.GLMResultsWrapper) -> None:
    FIGURES_DIR.mkdir(parents=True, exist_ok=True)

    fitted = model.fittedvalues
    resid = model.resid_pearson

    sample_size = min(10_000, len(resid))
    rng = np.random.default_rng(RANDOM_STATE)
    idx = rng.choice(np.arange(len(resid)), size=sample_size, replace=False)

    fitted_sample = fitted.iloc[idx]
    resid_sample = resid.iloc[idx]

    plt.figure(figsize=(8, 5))
    plt.scatter(fitted_sample, resid_sample, alpha=0.2)
    plt.axhline(0, linestyle="--")
    plt.title("Gamma GLM Residuals vs Fitted")
    plt.xlabel("Fitted values")
    plt.ylabel("Pearson residuals")
    plt.tight_layout()
    plt.savefig(FIGURES_DIR / "glm_residuals_vs_fitted_python.png", dpi=140)
    plt.close()

    plt.figure(figsize=(8, 5))
    stats.probplot(resid_sample, dist="norm", plot=plt)
    plt.title("Gamma GLM Normal Q-Q")
    plt.tight_layout()
    plt.savefig(FIGURES_DIR / "glm_qq_python.png", dpi=140)
    plt.close()


def main() -> None:
    print("Loading data from:", DATA_PATH)
    df = load_and_clean_data(DATA_PATH)

    print("\nRows, Columns:", df.shape)
    print("Missing values per column (top 10):")
    print(df.isna().sum().sort_values(ascending=False).head(10))
    print("Duplicate rows:", df.duplicated().sum())

    run_eda(df)
    run_inference(df)
    run_regression_workflow(df)

    print("\nDone. Figures saved in:", FIGURES_DIR)


if __name__ == "__main__":
    main()

File: notebooks/ed_length_of_stay_analysis.R
# Emergency Department Length of Stay: A Statistical Modeling Analysis
# Annelise Thorn
# STAT 5010: Statistical Methods and Applications II
# December 10, 2025

# ======================================================
# 1. Introduction, Research Question, and Techniques Used
# ======================================================

# This script is a direct R conversion of:
# notebooks/ed_length_of_stay_analysis.ipynb

# ========
# 2. Data
# ========

# 2.1 Data Import & Source
# Source: https://www.kaggle.com/datasets/aayushchou/hospital-length-of-stay-dataset-microsoft?resource=download

# load the dataset
df <- read.csv("data/raw/ed_length_of_stay_100k.csv")

# preview the dataset
head(df)

# 2.2 Data Description

# find number of rows and columns
dim(df)

# find names of columns
names(df)

# # summary statistics of the data
# summary(data)

# 2.3 Data Cleaning / Pre-processing

# check for missing values
colSums(is.na(df))

# check for duplicate rows
sum(duplicated(df))

# # check initial data types of each column
# str(df)

# # give the 5 number summary of numeric columns
# summary(df)

# convert date columns to Date type
df$vdate      <- as.Date(df$vdate, format = "%m/%d/%Y")
df$discharged <- as.Date(df$discharged, format = "%m/%d/%Y")

# convert gender to a factor
df$gender <- as.factor(df$gender)

# convert rcount to a factor and make ordered
df$rcount <- factor(
  df$rcount,
  levels  = c("0", "1", "2", "3", "4", "5+"),
  ordered = TRUE
)

# make facility ID a factor
df$facid <- as.factor(df$facid)

# make length of stay numeric
df$lengthofstay <- as.numeric(df$lengthofstay)

# remove impossible physiologic values
df <- df[df$sodium >= 110 & df$sodium <= 170, ]
df <- df[df$neutrophils >= 0 & df$neutrophils <= 100, ]
df <- df[df$glucose >= 0, ]
df <- df[df$bloodureanitro >= 0 & df$bloodureanitro <= 200, ]
df <- df[df$respiration >= 4 & df$respiration <= 40, ]
df <- df[df$hematocrit >= 5 & df$hematocrit <= 60, ]

# check for changes
str(df)

# ================================
# 3. Exploratory Data Analysis
# ================================

# check the 5 number summary of numeric columns
summary(df)

# # make boxplots for numeric columns to check for outliers
# numeric_columns <- sapply(df, is.numeric)
# for (col in names(df)[numeric_columns]) {
#   boxplot(df[[col]], main = paste("Boxplot of", col))
# }

# # make histograms of numeric / continuous variables
# hist(df$hemo,           breaks = 20, main = "Histogram of Hemoglobin")
# hist(df$hematocrit,     breaks = 20, main = "Histogram of Hematocrit")
# hist(df$neutrophils,    breaks = 20, main = "Histogram of Neutrophils")
# hist(df$sodium,         breaks = 20, main = "Histogram of Sodium")
# hist(df$glucose,        breaks = 20, main = "Histogram of Glucose")
# hist(df$bloodureanitro, breaks = 20, main = "Histogram of Blood Urea Nitrogen")
# hist(df$creatinine,     breaks = 20, main = "Histogram of Creatinine")
# hist(df$bmi,            breaks = 20, main = "Histogram of BMI")
# hist(df$pulse,          breaks = 20, main = "Histogram of Pulse Rate")
# hist(df$respiration,    breaks = 20, main = "Histogram of Respiration Rate")
hist(df$lengthofstay, breaks = 20, main = "Histogram of Length of Stay")

# # make barplots of binary / indicator INT variables
# barplot(table(df$dialysisrenalendstage), main = "Dialysis / Renal End-Stage")
# barplot(table(df$asthma),                main = "Asthma")
# barplot(table(df$irondef),               main = "Iron Deficiency")
# barplot(table(df$pneum),                 main = "Pneumonia")
# barplot(table(df$substancedependence),   main = "Substance Dependence")
# barplot(table(df$psychologicaldisordermajor), main = "Major Psychological Disorder")
# barplot(table(df$depress),               main = "Depression")
# barplot(table(df$psychother),            main = "Other Psychological Disorder")
# barplot(table(df$fibrosisandother),      main = "Fibrosis and Other")
# barplot(table(df$malnutrition),          main = "Malnutrition")

# # make barplot for other discrete INT variables
# barplot(table(df$secondarydiagnosisnonicd9),
#         main = "Secondary Diagnosis (non-ICD9) Count")

# # make barplots for categorical / factor variables
barplot(table(df$gender), main = "Gender Distribution")
# barplot(table(df$rcount), main = "Readmission Count")
barplot(table(df$facid),  main = "Facility Distribution")

# # make histograms for date variables
# hist(df$vdate,      breaks = "months", main = "Histogram of Visit Dates")
# hist(df$discharged, breaks = "months", main = "Histogram of Discharge Dates")

# =======================
# 4. Modeling & Inference
# =======================

# 4.1 Hypothesis Tests/Confidence Intervals

# t-test: Does LOS differ by gender?
# h0: There is no difference in the mean length of stay between male and female patients.
# h1: There is a difference in mean length of stay between males and females.

# two-sample t-test (two-sided)
ttest_gender <- t.test(lengthofstay ~ gender, data = df)

ttest_gender

# ANOVA: Does LOS differ across facilities?
# h0: All facilities have the same population mean length of stay.
# h1: At least one facility has a mean LOS that is different from the others.

# fit ANOVA model
anova_fac <- aov(lengthofstay ~ facid, data = df)

# ANOVA table
summary(anova_fac)

# Tukey test
TukeyHSD(anova_fac)

# 4.2 Baseline Regression Model (MLR)

# model with clinically relevant predictors
lm_mlr <- lm(
  lengthofstay ~ gender + rcount + facid +
    dialysisrenalendstage + asthma + irondef + pneum +
    substancedependence + psychologicaldisordermajor +
    depress + psychother + fibrosisandother +
    malnutrition + hemo +
    hematocrit + neutrophils + sodium + glucose +
    bloodureanitro + creatinine + bmi + pulse +
    respiration + secondarydiagnosisnonicd9,
  data = df
)

# model summary
summary(lm_mlr)

# 95% CIs for key coefficients
confint(lm_mlr)

# R-squared
summary(lm_mlr)$r.squared

# 4.3 Diagnostics for Baseline Model

# residual standard error
sigma(lm_mlr)

# information criteria for model comparison
AIC(lm_mlr)
BIC(lm_mlr)

par(mfrow = c(1, 1))

# residuals vs fitted
plot(lm_mlr, which = 1)

# normal Q-Q
plot(lm_mlr, which = 2)

# 4.4 Model Selection (AIC, BIC, MSPE)

# use the model as "full" model
form_full <- formula(lm_mlr)

# make more clinically-focused (drop some less important variables)
form_clinical <- lengthofstay ~
  rcount +
  dialysisrenalendstage + pneum +
  psychologicaldisordermajor + depress + malnutrition +
  hematocrit + neutrophils + sodium + glucose +
  bloodureanitro + creatinine + bmi + pulse + respiration

# "minimal" model using just a few strong predictors
form_minimal <- lengthofstay ~
  rcount + hematocrit + bloodureanitro + creatinine

# fit the three models on the full dataset
m_full     <- lm(form_full, data = df)
m_clinical <- lm(form_clinical, data = df)
m_minimal  <- lm(form_minimal, data = df)

# compare AIC and BIC
AIC(m_full, m_clinical, m_minimal)
BIC(m_full, m_clinical, m_minimal)

# train/test split and MSPE for the models
set.seed(123)

n <- nrow(df)
idx_train <- sample(seq_len(n), size = floor(0.7 * n))
train <- df[idx_train, ]
test  <- df[-idx_train, ]

# refit on training data
m_full_tr     <- lm(form_full, data = train)
m_clinical_tr <- lm(form_clinical, data = train)
m_minimal_tr  <- lm(form_minimal, data = train)

# true outcomes in test set
y_test <- test$lengthofstay

# predictions
pred_full     <- predict(m_full_tr, newdata = test)
pred_clinical <- predict(m_clinical_tr, newdata = test)
pred_minimal  <- predict(m_minimal_tr, newdata = test)

# MSPE for each model
mspe_results <- data.frame(
  Model = c("Full", "Clinical", "Minimal"),
  MSPE  = c(
    mean((y_test - pred_full)^2),
    mean((y_test - pred_clinical)^2),
    mean((y_test - pred_minimal)^2)
  )
)

mspe_results

# 4.5 Improved Model (GLM)

# use the clinically-motivated predictor set from 4.4
form_glm <- form_clinical

# fit Gamma GLM on the full dataset
glm_improved <- glm(
  formula = form_glm,
  family  = Gamma(link = "log"),
  data    = df
)

# model summary, AIC, and BIC
summary(glm_improved)
AIC(glm_improved)
BIC(glm_improved)

# Pseudo R^2 = 1 - (residual deviance / null deviance)
pseudo_R2_glm <- 1 - glm_improved$deviance / glm_improved$null.deviance
pseudo_R2_glm

# MSPE for the Gamma GLM

# fit the Gamma GLM on the training data
glm_improved_tr <- glm(
  formula = form_glm,
  family  = Gamma(link = "log"),
  data    = train
)

# predict LOS on the test set
pred_glm_improved <- predict(
  glm_improved_tr,
  newdata = test,
  type   = "response"
)

# MSPE
mspe_glm_improved <- mean((y_test - pred_glm_improved)^2)
mspe_glm_improved

# Add to the MSPE comparison table from 4.4
mspe_results_4_5 <- rbind(
  mspe_results,
  data.frame(
    Model = "Gamma GLM (improved)",
    MSPE  = mspe_glm_improved
  )
)

mspe_results_4_5

# 4.6 Final Model Diagnostics

# using the improved Gamma GLM as the final model
final_model <- glm_improved

# get fitted values and Pearson residuals
fitted_full <- fitted(final_model)
resid_full  <- residuals(final_model, type = "pearson")

# downsample residuals for plotting (large plots were causing the kernel to restart)
set.seed(123)
n       <- length(resid_full)
n_plot  <- min(10000, n)  # keep plots manageable
idx     <- sample(seq_len(n), size = n_plot)

fitted_resid <- fitted_full[idx]
resid_resid  <- resid_full[idx]

# basic summaries (full residuals)
summary(fitted_full)
summary(resid_full)

# diagnostic plots
op <- par(mfrow = c(2, 2))

# 1) Residuals vs fitted
plot(
  fitted_resid,
  resid_resid,
  xlab = "Fitted values",
  ylab = "Pearson residuals",
  main = "Residuals vs Fitted (Gamma GLM)"
)
abline(h = 0, lty = 2)

# 2) Normal Q-Q plot
qqnorm(resid_resid, main = "Normal Q-Q Plot (Gamma GLM)")
qqline(resid_resid, lty = 2)

# 3) Scale–location plot
plot(
  fitted_resid,
  sqrt(abs(resid_resid)),
  xlab = "Fitted values",
  ylab = "√|Pearson residuals|",
  main = "Scale–Location Plot"
)

# 4) Residuals vs index
plot(
  seq_along(resid_resid),
  resid_resid,
  xlab = "Index",
  ylab = "Pearson residuals",
  main = "Residuals vs Index"
)
abline(h = 0, lty = 2)

par(op)

# ============
# 5. Conclusion
# ============
# See notebook markdown for full narrative conclusion.

# ============
# 6. References
# ============
# https://www.kaggle.com/datasets/aayushchou/hospital-length-of-stay-dataset-microsoft?resource=download
# https://www.sciencedirect.com/science/article/pii/S1755599X20301026
